metadata/src/main/java/org/apache/kafka/controller/PartitionReassignmentRevert.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.controller;
+
+import org.apache.kafka.metadata.PartitionRegistration;
+import org.apache.kafka.metadata.Replicas;
+
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.List;
+import java.util.Objects;
+
+
+class PartitionReassignmentRevert {
+    private final List<Integer> replicas;
+    private final List<Integer> isr;
+
+    PartitionReassignmentRevert(PartitionRegistration registration) {
+        // Figure out the replica list and ISR that we will have after reverting the
+        // reassignment. In general, we want to take out any replica that the reassignment
+        // was adding, but keep the ones the reassignment was removing. (But see the
+        // special case below.)
+        Set<Integer> adding = Replicas.toSet(registration.addingReplicas);
+        this.replicas = new ArrayList<>(registration.replicas.length);
+        this.isr = new ArrayList<>(registration.isr.length);
+        for (int i = 0; i < registration.isr.length; i++) {
+            int replica = registration.isr[i];
+            if (!adding.contains(replica)) {
+                this.isr.add(replica);
+            } else if (i == registration.isr.length - 1 && isr.isEmpty()) {
+                // This is a special case where taking out all the "adding" replicas is
+                // not possible. The reason it is not possible is that doing so would
+                // create an empty ISR, which is not allowed.
+                //
+                // In this case, we leave in one of the adding replicas permanently.
```

Comment:
It's not possible to revert to the original replicas in all cases, though. Specifically, if none of the original replicas are in the ISR, you can't do such a revert without data loss. If the old controller is doing that unconditionally, it seems like a bug.

I suppose we could also fail the revert operation, but that doesn't seem like it would be a useful behavior from the user's point of view....

Reply:
One weird side effect of doing this is that it can change the replication factor for a partition to be different from the original one.

Reply:
I guess we can just not allow the reassignment to be cancelled in this situation. It should still be possible to undo the reassignment by creating a new reassignment, so it doesn't totally trap the user, I suppose...